parity-clib/src/java.rs

```diff
@@ -139,26 +138,10 @@ pub unsafe extern "system" fn Java_io_parity_ethereum_Parity_rpcQueryNative(
 	callback: JObject,
 	)
 {
-	let _ = async_checker(parity, rpc, callback, &env)
+	let _ = java_query_checker(parity, rpc, callback, &env)
 		.map(|(client, query, jvm, global_ref)| {
-			let callback = Arc::new(Callback::new(jvm, global_ref));
-			let cb = callback.clone();
-			let future = client.rpc_query(&query, None).map(move |response| {
-				let response = response.unwrap_or_else(|| error::EMPTY.to_string());
-				callback.call(&response);
-			});
-
-			let _handle = thread::Builder::new()
-				.name("rpc_query".to_string())
-				.spawn(move || {
-					let mut current_thread = CurrentThread::new();
-					current_thread.spawn(future);
-					let _ = current_thread.run_timeout(Duration::from_millis(timeout_ms as u64))
-						.map_err(|_e| {
-							cb.call(error::TIMEOUT);
-						});
-				})
-				.expect("rpc-query thread shouldn't fail; qed");
+			let callback = Arc::new(JavaCallback::new(jvm, global_ref));
+			parity_rpc_worker(client, &query, callback, timeout_ms as u64);
```

Comment:
We still spawn a thread per call, correct?

Reply:
Correct, I opened a ticket for this so we don't forget: https://github.com/paritytech/parity-ethereum/issues/10310